src/cartservice/cartstore/InstrumentedCartStore.cs

```diff
@@ -0,0 +1,116 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Threading.Tasks;
+using cartservice.interfaces;
+using Hipstershop;
+using OpenCensus.Trace;
+using OpenCensus.Trace.Sampler;
+
+namespace cartservice.cartstore
+{
+    /// <summary>
+    /// Wrapper for Cart Store - instrumented with OpenCensus tracing
+    /// </summary>
+    internal class InstrumentedCartStore : ICartStore
+    {
+        private readonly ICartStore cartStore;
+        private static ITracer tracer = Tracing.Tracer;
+        private ISpanBuilder initializeSpanBuilder, addItemSpanBuilder, emptySpanBuilder, getItemSpanBuilder, pingSpanBuilder; 
+
+        public InstrumentedCartStore(ICartStore cartStore)
+        {
+            this.cartStore = cartStore;
+
+            // Create Span Builders for tracing
+            initializeSpanBuilder = CreateSpanBuilder("Initialize Cart Store");
+            addItemSpanBuilder = CreateSpanBuilder("Add Item");
+            emptySpanBuilder = CreateSpanBuilder("Empty Cart");
+            getItemSpanBuilder = CreateSpanBuilder("Get Cart");
+            pingSpanBuilder = CreateSpanBuilder("Ping");
+        }
+        public Task AddItemAsync(string userId, string productId, int quantity)
+        {
+            using (var span = addItemSpanBuilder.StartScopedSpan())
+            {
+                return cartStore.AddItemAsync(userId, productId, quantity);
+            }
+        }
+
+        public Task EmptyCartAsync(string userId)
+        {
+            using (var span = emptySpanBuilder.StartScopedSpan())
+            {
+                return cartStore.EmptyCartAsync(userId);
+            }
+        }
+
+        public Task<Cart> GetCartAsync(string userId)
+        {
+            using (var span = getItemSpanBuilder.StartScopedSpan())
+            {
+                return cartStore.GetCartAsync(userId);
+            }
+        }
+
+        public Task InitializeAsync()
+        {
+            using (var span = initializeSpanBuilder.StartScopedSpan())
+            {
+                return cartStore.InitializeAsync();
+            }
+        }
+
+        public bool Ping()
+        {
+            using (var span = pingSpanBuilder.StartScopedSpan())
+            {
+                return cartStore.Ping();
+            }
+        }
+
+        public static ICartStore Create(string redis)
+        {
+            ICartStore cartStore;
+
+            // Redis was specified
+            if (!string.IsNullOrEmpty(redis))
```

Comment:
I think we should perform this `if` outside `ICartStore`.

This sounds a bit like code smell to me. Doing

```
if redis == ""
  return new LocalStore()
else
  return new RedisStore(redis)
```

is way better than

```
interface Store {

  method Create(redis) {
    if redis == ""
      return new LocalStore()
    else
      return new RedisStore(redis)
  }
}
```

because you just made the CartStore interface aware of its storage method (i.e. redis).

So please revert this logic to what it was before (first snippet above).

Reply:
I guess it's fine and still works the same way, so no need to change.